include/dlaf/permutations/general/impl.h

```diff
@@ -152,4 +157,345 @@ void Permutations<B, D, T, C>::call(SizeType i_begin, SizeType i_end, Matrix<con
                                                false>(dlaf::internal::Policy<B>(), std::move(permute_fn),
                                                       std::move(sender)));
 }
+
+template <class T, Device D>
+auto whenAllReadWriteTilesArray(LocalTileIndex begin, LocalTileIndex end, Matrix<T, D>& matrix) {
+  LocalTileSize sz{end.row() - begin.row() + 1, end.col() - begin.col() + 1};
+  namespace ex = pika::execution::experimental;
+  namespace ut = matrix::util;
+  return ex::when_all_vector(ut::collectReadWriteTiles(begin, sz, matrix));
+}
+
+template <class T, Device D>
+auto whenAllReadWriteTilesArray(Matrix<T, D>& matrix) {
+  namespace ex = pika::execution::experimental;
+  namespace ut = matrix::util;
+  return ex::when_all_vector(
+      ut::collectReadWriteTiles(LocalTileIndex(0, 0), matrix.distribution().localNrTiles(), matrix));
+}
+
+template <class T, Device D>
+auto whenAllReadOnlyTilesArray(LocalTileIndex begin, LocalTileIndex end, Matrix<const T, D>& matrix) {
+  LocalTileSize sz{end.row() - begin.row() + 1, end.col() - begin.col() + 1};
+  namespace ex = pika::execution::experimental;
+  namespace ut = matrix::util;
+  return ex::when_all_vector(ut::collectReadTiles(begin, sz, matrix));
+}
+
+template <class T, Device D>
+auto whenAllReadOnlyTilesArray(Matrix<const T, D>& matrix) {
+  namespace ex = pika::execution::experimental;
+  namespace ut = matrix::util;
+  return ex::when_all_vector(
+      ut::collectReadTiles(LocalTileIndex(0, 0), matrix.distribution().localNrTiles(), matrix));
+}
+
+// No data is sent or received but the processes participates in the collective call
+template <class T>
+void all2allEmptyData(const comm::Communicator& comm) {
+  namespace ex = pika::execution::experimental;
+
+  auto all2all_f = [comm](const std::vector<int>& arr, MPI_Request* req) {
+    MPI_Datatype dtype = dlaf::comm::mpi_datatype<std::remove_pointer_t<T>>::type;
+    // Avoid buffer aliasing error reported when send_ptr and recv_ptr are the same.
+    int irrelevant = 42;
+    DLAF_MPI_CHECK_ERROR(MPI_Ialltoallv(nullptr, arr.data(), arr.data(), dtype, &irrelevant, arr.data(),
+                                        arr.data(), dtype, comm, req));
+  };
+  dlaf::comm::internal::transformMPIDetach(std::move(all2all_f),
+                                           ex::just(std::vector<int>(to_sizet(comm.size()), 0)));
+}
+
+// Note: matrices are assumed to be in column-major layout!
+//
```

Comment:
The function `all2allData` is an internal detail, do you think it's worth adding asserts though? I improved the comment here https://github.com/eth-cscs/DLA-Future/pull/676/commits/cd18cd0eaab4f27947a16ad6a923ee522a2c0b0e .

Reply:
It depends on how much work the asserts would be! :wink: And how silently/badly it would fail elsewhere if that assumption is broken somewhere. But this is not a blocker for merging.